### ISSUE TITLE
Rule proposal: ensure exports satisfy certain types

### ISSUE THREAD

**Issue by turadg** [typescript-eslint/typescript-eslint#9599]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Configuration can name expected exports and what type they must be assignable to. (i.e. satisfy)

A wildcard '*' name can be used to enforce all unspecified exports satisfy a particular type.

### Fail Cases

```TypeScript
# options: [{ 'start': 'StartFn' }],
export type StartFn = (name: string) => void;
export const start = (num: number) => {};

# options: [{ '*': 'number' }],
export const SECOND_AS_MS = 1000;
export const MINUTE_AS_MS = 60 * SECOND_AS_MS;
export const DAY_AS_MS = 24 * HOUR_AS_MS;
```

### Pass Cases

```TypeScript
# options: [{ start: 'StartFn', '*': 'MyType' }],

export type StartFn = (name: string) => void;
export const start: StartFn = (name: string) => {};

export type MyType = { name: string };
export const other: MyType = null;
```

### Additional Info

Working prototype: https://github.com/turadg/typescript-eslint/tree/ta/exports-satisfy

**Comment by bradzacher** [comment]
> I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

I'm not sure how this rule passes this check.
The desire to make sure a particular set of files adheres to a specific shape of export is a very, very niche requirement - it's rare that codebase will actively want to do this.

For this reason I'm going to reject this proposal for this project.

**Comment by Josh-Cena** [comment]
I can clearly see the value for this rule, but I agree it's not fit in our general-purpose linter, because it neither enforces a consistent style nor catches potential mistakes. Even in a single code base, this is unlikely to be applied globally, only to a small set of files, so I think you should build it yourself, or find an existing one.
